api/v1alpha1/inclusterippool_conversion.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2023 Deutsche Telekom AG.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//nolint:forcetypeassert,golint,revive,stylecheck
+package v1alpha1
+
+import (
+	utilconversion "sigs.k8s.io/cluster-api/util/conversion"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+
+	v1alpha2 "github.com/telekom/cluster-api-ipam-provider-in-cluster/api/v1alpha2"
+)
+
+// ConvertTo converts v1alpha1.InClusterIPPool to v1alpha2.InClusterIPPool.
+func (src *InClusterIPPool) ConvertTo(dstRaw conversion.Hub) error {
+	dst := dstRaw.(*v1alpha2.InClusterIPPool)
+	if err := Convert_v1alpha1_InClusterIPPool_To_v1alpha2_InClusterIPPool(src, dst, nil); err != nil {
+		return err
+	}
+
+	if err := utilconversion.MarshalData(src, dst); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ConvertTo converts v1alpha2.InClusterIPPool to v1alpha1.InClusterIPPool.
+func (dst *InClusterIPPool) ConvertFrom(srcRaw conversion.Hub) error {
+	src := srcRaw.(*v1alpha2.InClusterIPPool)
+	if err := Convert_v1alpha2_InClusterIPPool_To_v1alpha1_InClusterIPPool(src, dst, nil); err != nil {
+		return err
+	}
+
+	restored := &InClusterIPPool{}
+	if ok, err := utilconversion.UnmarshalData(src, restored); err != nil || !ok {
+		return err
+	}
+
+	dst.Spec = restored.Spec
```

Comment:
I don't think this will work. It completely ignores any modifications made to the resource using v1alpha2, doesn't it?

IF we want to do proper api versioning, and want to implement conversion, I think it needs to be a bit more sophisticated. It should be possible to properly convert between the versions, since no information is removed. The layout just changes, becomes narrower.
`.spec.first` and `.spec.last` can only be used when a single range of addresses is specified. Same goes for subnet. So we should only fill those values when converting alpha2->alpha1 when `.spec.addresses` contains only one entry. Otherwise we'll just use `.spec.addresses`.
The only issue is: we need to know which format was used with alpha1. We can simply store that in an annotation (only if it's first/last or subnet - addresses is the default) when converting 1->2 and then _try_ to respect that annotation when converting 2->1.

Generally adding annotations to the hub version isn't very pretty, since those will remain even if conversion is removed at some point, but there probably is no better way to handle it here.

Reply:
I suggest we just go with using `.addresses` for 2->1. It feels like a decent effort to get the conversion back to `.first` and `.last` correct and I'm not sure what it buys us other than some presentation to a user. The behavior is the same regardless of which field is used to represent the data. Any concerns that you have with us going ahead with that approach?

Reply:
I was thinking about automated tooling that's still consuming alpha1. Tools like flux would fetch the resource, compare it to their known state and compare. If it deviates, it would apply again. Or tooling that applies and then checks whether it's correct. If we just use `.addresses`, what you write wouldn't be what you get back from the api server.

On the other hand, it might not be worth the annotation that we'll have to clean up at some point in the future (not too big of a deal though).

Reply:
Thinking about it more deeply with Aidan and team. If we want to support returning the values they had in v1 in the same format they inputted the data in v1 we should only do that if they last updated the data in v1 i.e if they initially apply in v1, and make updates in v1, we should return what they had in v1. If they apply in v1, but make an update in v2, we should start returning it as v2 instead of figuring out what they initially did in v1 and transforming v2 data.

Thinking about what this might be implemented as. Ideally we could remove the annotation if they make an update in v2, but not sure if there is a way to do that cleanly. I was thinking it could be removed in the webhook, but v1 is already converted to v2 at that point.

The other implementation would be to conditionally apply the `restored.Spec` if the up-converted restored spec matches the current spec (this means the annotation is still valid and therefore we can return it).

Reply:
We could just remove the annotation once we remove v1alpha1 entirely. It doesn't really matter whether it's there or not.

I didn't get the last paragraph. How would you check that?

Reply:
The thought was we could take what was in the annotation and convert it to v1alpha2 to compare it apples to apples with the current spec. If it matches we use what was in the annotation and if it doesn't we return what is in the current spec. To me this is okay because if they were to make an update using v1alpha2 I'm assuming then they wouldn't want to downconvert to use the first and last format.

And because this is somewhat confusing to talk about here is psuedo code on what I'm thinking:

```go
restoredV2 := &v1alpha2.InClusterIPPool{}
if err := Convert_v1alpha1_InClusterIPPool_To_v1alpha2_InClusterIPPool(restored, restoredV2, nil); err != nil {
		return err
}

if restoredV2.spec.addresses == src.spec.addresses &&
		restoredV2.spec.prefix == src.spec.prefix &&
		restoredV2.spec.gateway == src.spec.gateway {
		dst.Spec = restored.Spec
}
```

Reply:
Ah, now it makes sense to me. Yes, we can also go with that approach. I'll let you decide.

Reply:
Okay, I've updated the v2alpha1 to v1alpha1 convert to do the above approach. Thanks.